src/common/Formatter.cc

```diff
@@ -544,6 +546,13 @@ void XMLFormatter::write_raw_data(const char *data)
   m_ss << data;
 }
 
+void XMLFormatter::write_bin_data(const char *buff, int buf_len)
+{
+  std::stringbuf *pbuf = m_ss.rdbuf();
```

Comment:
I'll stop commenting on these;  basically, we should have "T* foo"  and "T& foo" for all values of T